### PR TITLE
EIP-3 (BIP-44) implementation 

### PR DESCRIPTION
**Pull request by kushti**
This PR is about EIP-3 implementation ( see issue #959 ). For the sake of compatibility with the wallets made with previous version sand miner, following rules apply, based on an observation that users import old wallets via copying files (/wallet folder in case of 3.3.x, /wallet/keystore and /wallet/storage in case of 3.2.x) or /wallet/restore. Please also note that older versions expose public key corresponding to the master key, while new version hides it (while maintaining the master key and using it in the scans) when it is possible.

Thus derivation works as follows:
* if no keys derived yet, derive it against m/44'/429'/0'/0/0 by default, or against m/1 if ergo.wallet.preEip3Derivation = true (false by default)
* if there're keys in the wallet, derive by using old scheme (m/1 -> m/2, m/44'/429'/0'/0/0 -> m/44'/429'/0'/1) if last key is not in the EIP-3 space (not starting with m/44'/429'/0'), otherwise, derive via new procedure (m/44'/429'/0'/0/0 -> m/44'/429'/0'/0/1, note that the result if different from the old scheme).

Thus in case of import from files everything should be ok, in case of restoring a wallet from mnemonic, funds will be found automatically if there were no derived keys there, otherwise, preEip3Derivation flag is needed to be set. 

Also:  
* hide a public key corresponding to the master key if derivation path of the first derived key starts with m/44'/429'/0' in /wallet/addresses output , the key is participating in scans still
* change address by default is set to m/44'/429'/0'/0/0 
* send public key corresponding to master key to the miners , so miningScript (/mining/rewardsAddress) is the same as before.

Close: #959 , #1082

### CODE REVIEW

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala [hunk 1]
```diff
@@ -41,6 +50,13 @@ final case class DerivationPath(decodedPath: Seq[Int], publicBranch: Boolean) {
     */
   def toPrivateBranch: DerivationPath = this.copy(publicBranch = false)
 
+  /**
+    * @return whether derivation path corresponds to EIP-3
+    */
+  def isEip3: Boolean = {
```
Comment (greenhat) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala:56]: Should we consider to encode pre-EIP3 vs. EIP3 paths on a type level (ADT, etc.)?

**Review by greenhat — CHANGES_REQUESTED** [review]
Looks good, please, see the comments.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/Constants.scala [hunk 2]
```diff
@@ -25,4 +26,28 @@ object Constants {
...
+    * Following this idea we should use next scheme
+    *
+    * m / 44' / 429' / account' / change / address_index
+    */
+  val CoinType = 429
+
+  /**
+    * Pre - EIP3 derivation path
+    */
+  val oldDerivation: DerivationPath = DerivationPath.fromEncoded("m/1").get
```
[the remark this answers is not in the document]
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/Constants.scala]: done

File: src/main/scala/org/ergoplatform/settings/WalletSettings.scala [hunk 3]
```diff
@@ -6,6 +6,7 @@ import org.ergoplatform.wallet.settings.SecretStorageSettings
 case class WalletSettings(secretStorage: SecretStorageSettings,
                           seedStrengthBits: Int,
                           mnemonicPhraseLanguage: String,
+                          oldDerivation: Boolean = false,
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/settings/WalletSettings.scala]: done (but preEip3Derivation , not *Path)

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala [hunk 4]
```diff
@@ -71,8 +87,11 @@ object DerivationPath {
 
   /**
     * Finds next available path index for a new key.
+    * @param secrets - secrets previously generated
+    * @param oldDerivation - whether to use pre-EIP3 derivation or not
     */
-  def nextPath(secrets: IndexedSeq[ExtendedSecretKey]): Try[DerivationPath] = {
+  def nextPath(secrets: IndexedSeq[ExtendedSecretKey],
+               oldDerivation: Boolean = false): Try[DerivationPath] = {
```
[the remark this answers is not in the document]
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala]: done (but preEip3Derivation , not *Path)

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala (hunk 1 above)
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala:56]: I dunno, maybe not so critical difference to disinguish on the type level?

**Comment by kushti** [comment]
@greenhat addressed comments, please make another pass!

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala (hunk 1 above)
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala:56]: Yeah, I’m not completely sure here. Seems useful to further distinguish code paths that handle old and new derivation. Do they intended to co-exist forever or is it more like a “migration” kind of scenario?

**Review by greenhat — APPROVED** [review]

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala [hunk 5]
```diff
@@ -19,4 +22,105 @@ class DerivationPathSpec
...
+    implicit val enc = new ErgoAddressEncoder(ErgoAddressEncoder.MainnetNetworkPrefix)
+
+    // This testing pair is checked against CoinBarn and Yoroi
+    val mnemonic = "liar exercise solve delay betray sheriff method empower disease river recall vacuum"
+    val address = "9hAymcGaRfTX7bMADNdfWfk7CKzi2ZpvRBCmtEf6d92n8E26Ax7"
+
+    val seed = Mnemonic.toSeed(mnemonic, None)
+
+    val masterKey = ExtendedSecretKey.deriveMasterKey(seed)
+    val dp = DerivationPath.nextPath(IndexedSeq(masterKey)).get
```
Comment (aslesarenko) [inline ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala:35]:
```suggestion
    val dp = DerivationPath.nextPath(IndexedSeq(masterKey), usePreEip3Derivation = false).get
```

Comment (aslesarenko) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala:35]: Suggest to pass this explicitly for readability

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala [hunk 6]
```diff
@@ -19,4 +22,105 @@ class DerivationPathSpec
...
+
+    val seed = Mnemonic.toSeed(mnemonic, None)
+
+    val masterKey = ExtendedSecretKey.deriveMasterKey(seed)
+    val dp = DerivationPath.nextPath(IndexedSeq(masterKey), usePreEip3Derivation = true).get
+    val sk = masterKey.derive(dp).asInstanceOf[ExtendedSecretKey]
+    val pk = sk.publicKey.key
+    P2PKAddress(pk).toString() shouldBe address
+
+    val dp2 = DerivationPath.nextPath(IndexedSeq(masterKey, sk)).get
```
Comment (aslesarenko) [inline ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala:76]: So, the usePreEip3Derivation = false is ignored in this case?

Comment (aslesarenko) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala:76]: Please describe the behavior for this test case in ScalaDoc of the method.

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala [hunk 7]
```diff
@@ -67,20 +67,29 @@ object WalletCache {
...
 
-  def trackedAddresses(trackedPubKeys: Seq[ExtendedPublicKey],
-                       addressEncoder: ErgoAddressEncoder): Seq[P2PKAddress] =
-    trackedPubKeys.map(pk => P2PKAddress(pk.key)(addressEncoder))
+  private def publicKeyAddresses(trackedPubKeys: Seq[ExtendedPublicKey],
+                                 addressEncoder: ErgoAddressEncoder): Seq[P2PKAddress] = {
+    // remove master key from visible addresses if following-up keys are from EIP-3 range
+    val pks = if (trackedPubKeys.size > 1
+      && trackedPubKeys.head.path.isMaster
+      && trackedPubKeys(1).path.isEip3) {
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala:78]: Suggest to extract this condition into a named predicate.

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala [hunk 8]
```diff
@@ -67,20 +67,29 @@ object WalletCache {
...
+      trackedPubKeys.tail
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala:79]: Please add a clarifying comment, why `tail` is used here.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala (hunk 5 above)
Comment (kushti) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala]: done

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala (hunk 6 above)
Comment (kushti) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala]: @aslesarenko are we going to introduce ScalaDoc requirements for testing code ?

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala (hunk 7 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala:78]: But why? As it is used once in a small method (these lines take like a half of the method), then readability will not be improved I guess.

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala (hunk 8 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala:79]: Was done on line 75, add another comment though

**Comment by kushti** [comment]
@aslesarenko comments addressed, please make another pass

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala (hunk 1 above)
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/DerivationPath.scala:56]: @greenhat As there's (and likely) will be a way to derive from a custom path, the schemes will coexist. 

We do not follow BIP-44 fully, namely, we do not separate between internal and external addresses, maybe some support will come in future but hard to tell atm

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala (hunk 6 above)
Comment (aslesarenko) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala]: @kushti no, I just don't understand the code. 
As always, please don't spend time writing he response here, improve the code instead, if you will.

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala (hunk 7 above)
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala:78]: What this predicate means? Looks like some intermediate state of the WalletCache. If yes, why don't give it a name?

Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/WalletCache.scala:78]: @aslesarenko there's the comment (line 75) explaining what is checked in the code

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala (hunk 6 above)
Comment (kushti) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/secrets/DerivationPathSpec.scala]: @aslesarenko ScalaDoc is not needed then it seems, added comments.

**Comment by kushti** [comment]
@aslesarenko comments has been addressed, please make another pass!

**Review by aslesarenko — APPROVED** [review]

**Pull request merged**